proposed/disable-multi-level-lookup-by-default.md

```diff
@@ -0,0 +1,204 @@
+running applications since the app may depend on exact output format. As such
+the additional information should only be provided when running CLI commands (SDK).
+
+Host should print out a similar message when enumerating runtimes and SDKs in
+`dotnet --info`. As mentioned above `dotnet --list-runtimes` and `dotnet --list-sdks`
+should not be affected.
+
+The message should only be shown when there's observable difference in behavior.
+For CLI commands that means when running the command with MLL on-by-default
+would end up using different version of SDK then when running it with
+MLL off-by-default. `dotnet --info` should print the message in all cases where
+there's a global install location (on top of the private one which is used
+to run the command).
+
+## Open questions
+
+### New dotnet CLI actions for install/update
+
+This is effectively an extension to the discrepancy between app behavior and
+`dotnet --info` behavior. What is the desired behavior for the new CLI actions -
+this is yet to be determined.
+
+### Way to dismiss informational message about the change
+
+There should be some way to dismiss the information messages printed by
+CLI commands if they are affected.
+
+1. Require the user to set `DOTNET_MULTILEVEL_LOOKUP=0` - which will disable
+   the message.
+1. Only show the message once (would require a way to remember that it was shown,
+   similar to first run CLI experience).
+1. Introduce some other mechanism (new env. variable?)
+
+I personally don't like either option. 2 is too restrictive and might make
+the message effectively useless. 1 feels counterproductive as it basically goes
+back to current state where setting `DOTNET_MULTILEVEL_LOOKUP=0` is very common.
+3 feels even worse as it introduces yet another env. variable to set.
+
+### Hosting APIs accept install location as parameter
+
+Currently most hosting APIs take a parameter which specifies the "dotnet root",
+so that they can behave as if one used `dotnet` from that location.
+But the APIs themselves are implemented in `hostfxr` which can be of any version.
+This is already a niche problem (if a really old `hostfxr` is used to run things
+on a really new private install the behavior is a bit problematic).
+With this change such discrepancy would become more apparent.
+So the question is to which version should the APIs react to:
+
+1. Use the version of the `hostfxr` itself - this would be similar to the above
+   described behavior of reacting to the version of the `dotnet.exe` itself,
+   but it might be a bit unexpected.
+1. Use the version of the specified "dotnet root" - this would be more in line
+   with what the "dotnet root" parameter tries to convey. This might be technically
+   a bit challenging in some cases.
+
+I think it should be OK to go with option 1 - rely on the version of `hostfxr` -
+but this will need more detailed investigation.
+
+## Alternative solutions
+
+### Off by default everywhere
```

Comment:
We will effectively end up here eventually once .NET 5 and lower are out of support.

We may want to say that we will delete support multi-level-lookup one year after .NET 5 and lower are out of support, so that we do not have to carry it forever.

Reply:
https://github.com/dotnet/designs/pull/178/files#diff-1ae0e0e27bffb4488839ad85db6c7024d71b809120c6a4cf46fcae385ba3f745R51 would have to be changed to this to work.

Reply:
If we go with something like this, we could also show the warning message unconditionally in places like `dotnet --info`, and then just remove the message when we remove MLL in a future patch.

Reply:
Changed the proposal to remove MLL for .NET 7+ apps. Also added a section about the plan to completely remove MLL for everything once all downlevel runtimes are out of support.